robot/impl/local_robot.go

```diff
@@ -498,6 +502,8 @@ func newWithResources(
 	}
 
 	successful = true
+	// Robot is "initializing" until first reconfigure after initial creation completes.
+	r.initializing.Store(true)
```

Comment:
Technically there's a `Reconfigure` earlier on newcode 491 that sets this value to false. Not to mention the `initializing` value is initialized (ugh) to false. Two things:
* I'm taking that it's important we exit this function with `initializing` set to true. But I would expect to see the setting up at the top near the constructor. Can we document that the placement here is intentional to avoid prior calls mucking with the state?
* Is this function guaranteed to not start webserver and expose the `GetMachineStatus` API? If it can, it seems we might be setting `initializing` too late and may allow clients to observe an illegal transition of not initializing -> initializing.

Reply:
Also -- line [428 (old) 432 (new)](https://github.com/viamrobotics/rdk/pull/4616/files#diff-0b975da6011656cd3f2caf0435023573bdb5ac7c6a425e7917dd5cc9d6be63e4R432) refers to the mod manager web server. Do we need to consider/provide guidelines for how module SDKs (which are -- in theory -- different from "application SDKs") use and perhaps expose `GetMachineStatus`?

Reply:
I've modified the mechanics here slightly. You'll see there's a new `robot.Option` to start a robot in initializing mode. You can then use `SetInitializing(false)` to mark the robot as running. This means that only the code here in `web/server/entrypoint.go` is "special" with respect to initialization. All other calls to `robotimpl.New` will create robots that always return `robot.StateRunning` from `MachineStatus`. 

I'm not sure that will address all your concerns here, and I'll think a bit harder about your module question before re-requesting review.

Reply:
> Do we need to consider/provide guidelines for how module SDKs (which are -- in theory -- different from "application SDKs") use and perhaps expose GetMachineStatus?

Slightly confused about your question here. The mod manager web server is started before any modules are added to the module manager, so it should always be the case that "module SDKs" (I'm reading that as Golang, Python, and C++ module libraries) should have the ability to connect back to the RDK through the mod manager web server before any module process has started. Were you suggesting that the module libraries should be using `MachineStatus` to check the status of the parent RDK, or that they should expose their own `MachineStatus` endpoint for some reason?

Reply:
> Slightly confused about your question here. The mod manager web server is started before any modules are added to the module manager, so it should always be the case that "module SDKs" (I'm reading that as Golang, Python, and C++ module libraries) should have the ability to connect back to the RDK through the mod manager web server before any module process has started.

Given that, it sounds like if* a module, as soon as it was possible, tried calling `MachineStatus` they'd get an initialized == false.

But it also sounds like in the current, pre-patch code, a module could actually call `ResourceNames` before a regular "network client" could? Because we wouldn't have started accepting connections yet? And the result of calling `ResourceNames` would be undefined as the robot hasn't necessarily done/completed its initial reconfigure yet? Just considering the "happy path" where the initial robot config is good.

And if that's true, modules already have to be "resilient" to talking with an "uninitialized" robot. And we would not expect to need module changes. Such as the testing changes to "wait by default" for a robot to be initialized.

Reply:
> Given that, it sounds like if* a module, as soon as it was possible, tried calling MachineStatus they'd get an initialized == false.

That sounds correct to me.

> But it also sounds like in the current, pre-patch code, a module could actually call ResourceNames before a regular "network client" could? Because we wouldn't have started accepting connections yet? 

That sounds correct to me. In particular, the "module web server" of the RDK will be open for connections while the regular web server will not be. So, a module could connect via the former before a regular "network client" could connect via the latter.

> And the result of calling ResourceNames would be undefined as the robot hasn't necessarily done/completed its initial reconfigure yet?

That does not sound correct to me. If a module is able to call `ResourceNames` through the module server, it will see the current status of the resource graph in terms of available names. If some resources have already completed configuration, the module will see them through `ResourceNames`. 

> And if that's true, modules already have to be "resilient" to talking with an "uninitialized" robot. And we would not expect to need module changes. 

Modules do expect a certain guarantee around modular dependencies: if a modular resource A depends on another resource B, it is expected that, barring any inability to create B, B will be available and usable through `ResourceByName` within the constructor for A. I actually broke that guarantee and caused `TestComplexModule` to fail [here](https://github.com/viamrobotics/rdk/blob/main/examples/customresources/models/mybase/mybase.go#L78-L79). 

We need the web service, which is "weakly" (it's actually, annoyingly, a fourth type of hardcoded-weak dependency that is not registered via `WeakDependencies` in resource registration) dependent on all resources to `Reconfigure` before the modular base builds such that the web service is aware of the two motors that the modular base depends on.

I "fixed" the test by updating weak dependents more often and more simply in `completeConfig` (see my incoming comment.)